[mlir][vector] Canonicalize gathers/scatters with trivial offsets

Hardcode84 opened the pull request:
Canonicalize gathers/scatters with contiguous (i.e. [0, 1, 2, ...]) offsets into vector masked load/store ops.

Hardcode84 committed:
[mlir][vector] Canonicalize gathers/scatters with trivial offsets

Cononicalize gathers/scatters with contiguous (i.e. [0, 1, 2, ...]) offsets into vector masked load/store ops.

dcaballe commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5181,6 +5181,19 @@ std::optional<SmallVector<int64_t, 4>> GatherOp::getShapeForUnroll() {
   return llvm::to_vector<4>(getVectorType().getShape());
 }
 
+static LogicalResult isContiguousIndices(Value val) {
```
Could you add doc about the current supported cases and limitations.

nit: `val` -> `indices`, `indexVec` ... ?

dcaballe commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5181,6 +5181,19 @@ std::optional<SmallVector<int64_t, 4>> GatherOp::getShapeForUnroll() {
+    return failure();
+
+  DenseIntElementsAttr elements;
+  if (!matchPattern(val, m_Constant(&elements)))
```
We may want to do something for `ConstantMaskOp`

dcaballe commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5181,6 +5181,19 @@ std::optional<SmallVector<int64_t, 4>> GatherOp::getShapeForUnroll() {
+    return failure();
+
+  return success(
+      llvm::equal(elements, llvm::seq<int64_t>(0, vecType.getNumElements())));
```
What about contiguous indices with a different start number?

dcaballe commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5181,6 +5181,19 @@ std::optional<SmallVector<int64_t, 4>> GatherOp::getShapeForUnroll() {
+
+  return success(
+      llvm::equal(elements, llvm::seq<int64_t>(0, vecType.getNumElements())));
+}
```
@banach-space, is there a common utility that we can use here and for the extract op in the Linalg vectorizer?

banach-space replied:
Not yet  - the vectorizer looks at the scalar indices _before_ vectorization. However, this patch make me think that we could do better 🤔 Let me look into this!

banach-space replied:
@Hardcode84 Would you have some linalg examples that vectorize into these contiguous gathers? That would be helpful, but no worries if `Vector` is your actual starting point here.

dcaballe commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -2826,3 +2826,34 @@ func.func @contiguous_extract_strided_slices_to_extract_failure_non_full_inner_s
+func.func @contiguous_gather(%base: memref<?xf32>,
+                             %mask: vector<16xi1>, %passthru: vector<16xf32>) -> vector<16xf32> {
+  %c0 = arith.constant 0 : index
+  %indices = arith.constant dense<[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]> : vector<16xi32>
```
Could you add tests for:
  * Start index != 0
  * ConstantMaskOp
  * constant indices that describe a broadcast (e.g., `[3, 3, 3, 3, 3, 3... 3]`)

Hardcode84 replied:
* We don't need any special handling for constant mask as it already handled in existing masked -> non-masked canonicalizations, added a couple of tests.
* I can add support for non-zero start, but broadcast is more involved
  * For scatters duplicated indices are undefined per current spec
  * For gather we need `reduce(mask)` + 1-element `vector.maskedload` + `extract` + `splat` and I would rather not do this as part of this PR.

banach-space replied:
> broadcast is more involved

Could you add a test and a TODO? The test would be "negative" (i.e. the folder would leave the code unchanged). For "scatter" load we'd only need to make sure that invalid.mlir contains relevant test. Could you check that?

> I can add support for non-zero start

Could you add a negative test to exercise this case? And a TODO to extend the pattern :)

Hardcode84 replied:
Invalid scatter indices (and invalid dynamic indices in general) should not fail validation (see https://mlir.llvm.org/getting_started/DeveloperGuide/#ir-verifier), so nothing to add to `invalid.mlir`

dcaballe reviewed (commented):
Awesome, thanks! I took a quick look and left some comments for now.

banach-space reviewed (commented):
Great, thank you for working on this!

Two high level requests:
* add support for `vector.step`
* either add support for multi-dim gathers/scatters, or document the limitations.

Hardcode84 committed:
add const mask tests

Hardcode84 committed:
nits

Hardcode84 commented:
> add support for vector.step

Maybe, instead of adding special case for `vector.step` just fold all non-scalable `vector.step`s into `arith.constant`s? It will probably help other canonicalizations as well.

Hardcode84 committed:
vector.step support

dcaballe commented:
> Maybe, instead of adding special case for vector.step just fold all non-scalable vector.steps into arith.constants? It will probably help other canonicalizations as well.

I would say the other way around. We may want `vector.step` to be the canonical form for vector constants that fall into that category as `vector.step` holds valuable information that has to be inferred by inspecting the values of a constant op.

Hardcode84 commented:
Added `vector.step` support and rebased but I don't want to do non-zero start offset and broadcast as part of this PR.

> I would say the other way around. We may want vector.step to be the canonical form for vector constants that fall into that category as vector.step holds valuable information that has to be inferred by inspecting the values of a constant op.

If we really want `vector.step` to be a canonical representation, we will need to always canonicalize `[0,1,2...]` constants to it (and it to be part of `canonicalize` pass and not just some random set of patterns), otherwise relevant patterns like this one will need to always check both forms. I'm still not convinced `vector.step` (for non-scalable) if useful as checking for relevant constant pattern is trivial and actually gives you more freedom (e.g. you can check for non-zero start offsets mentioned earlier) but I will leave this fight for someone else.

Hardcode84 commented:
Actually, I think, better approach may be to have a special `StepElementsAttr` in the same way we already have `SplatElementsAttr`, which is "canonicalized" on `DenseElementsAttr` construction like `SplatElementsAttr`. This way we don't need any special ops and users which doesn't know or doesn't care about `StepElementsAttr` can still work with it as normal dense constant.

banach-space commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5202,11 +5219,28 @@ class GatherFolder final : public OpRewritePattern<GatherOp> {
   }
 };
+
+/// Fold gathers with consecutive offsets [0, 1, 2, ...] into contiguous
+/// maskedload. Only 1D non-scalable vectors are supported for now.
```
Please add tests for scalable vectors and highlight where in the implementation the assumption about "non-scalability" is made.  Also, please add a negative test with 2D vectors.

[nit] We tend to say "fixed width vs scalable" rather than "non-scalable vs scalable" :_

banach-space commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5202,11 +5219,28 @@ class GatherFolder final : public OpRewritePattern<GatherOp> {
+
+/// Fold gathers with consecutive offsets [0, 1, 2, ...] into contiguous
+/// maskedload. Only 1D non-scalable vectors are supported for now.
+class GatherTrivialIndices final : public OpRewritePattern<GatherOp> {
```
`GatherTrivialIndices` -> `FoldContiguousGather`?

banach-space commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -5184,6 +5184,23 @@ std::optional<SmallVector<int64_t, 4>> GatherOp::getShapeForUnroll() {
 }
 
+/// Cheeck if `indexVec` is constant 1D vec of consecutive values [0, 1, 2, ...]
+static LogicalResult isContiguousIndices(Value indexVec) {
```
This method doesn't care where the input came from (it could be indices, but it could be sth else). I suggest generalising the name.
```suggestion
static LogicalResult isZeroBasedContiguousSeq(Value indexVec) {
```

banach-space commented:
> Actually, I think, better approach may be to have a special StepElementsAttr in the same way we already have SplatElementsAttr, which is "canonicalized" on DenseElementsAttr construction like SplatElementsAttr. This way we don't need any special ops and users which doesn't know or doesn't care about StepElementsAttr can still work with it as normal dense constant.

Supporting `vector.step` is important - that's the only option for scalable vectors. In fact, since you added support for `vector.step`,  your changes should also work for scalable vectors.

Hardcode84 commented:
> @Hardcode84 Would you have some linalg examples that vectorize into these contiguous gathers? That would be helpful, but no worries if Vector is your actual starting point here.

In my specific case (https://github.com/iree-org/iree-turbine/blob/main/iree/turbine/kernel/wave/codegen.py#L808), I generate gathers directly, without going through linalg. Also as side note, indices are constructed from user-provided sympy exprs, so we don't know beforehand if it's a step or not.

Hardcode84 committed:
rename func

Hardcode84 committed:
more tests

Hardcode84 commented:
updated

Hardcode84 committed:
fix name

banach-space commented:
@Hardcode84 Is this ready for another round of reviews? If yes, could you follow LLVM's code-review guidelines and ping reviewers? 

* https://llvm.org/docs/CodeReview.html#code-reviews-speed-and-reciprocity

> Ping the patch. If it is urgent, provide reasons why it is important to you to get this patch landed and ping it every couple of days. **If it is not urgent, the common courtesy ping rate is one week**. Remember that you’re asking for valuable time from other professional developers.

Note, reviewers get a notification every time a PR is updated (and there's a lot of PRs). "ping" is the usual LLVM way let people know that this is ready for another round :) And, specifically, things like:

> updated

tend to interpreted as noise (it's not clear to me whether "updated" means "I've addressed all PR comments" or just casual "I've made _some_ changes, but might do some _more_ sometime soon.").

Thanks :)
-Andrzej

Hardcode84 commented:
@banach-space yes, PTAL.

I understand everyone is busy but this PR was intended as trivial improvement and I never planned for it to take multiple months or to support broadcasts/step/scalable vectors/non-1D vectors (all of which have zero benefit for my specific use case). At this point I would prefer to either merge it in current form or just drop it and move on.

banach-space reviewed (approved):
LGTM, thanks!

@dcaballe , looks like tests in canonicalize.mlir cover all TODOs (apart from scalable vectors, but I can handle that). WDYT?

Hardcode84 commented:
Thanks, sorry for the rant
